src/solvers/dgsem_tree/dg_2d.jl

```diff
@@ -490,6 +511,162 @@ end
+  calcflux_fhat!(fhat1, fhat2, u, mesh,
+      nonconservative_terms, equations, volume_flux_dg, dg, element, cache)
+
+  # low-order FV fluxes
+  @unpack fstar1_L_threaded, fstar1_R_threaded, fstar2_L_threaded, fstar2_R_threaded = cache
+
+  fstar1_L = fstar1_L_threaded[Threads.threadid()]
+  fstar2_L = fstar2_L_threaded[Threads.threadid()]
+  fstar1_R = fstar1_R_threaded[Threads.threadid()]
+  fstar2_R = fstar2_R_threaded[Threads.threadid()]
+  calcflux_fv!(fstar1_L, fstar1_R, fstar2_L, fstar2_R, u, mesh,
+      nonconservative_terms, equations, volume_flux_fv, dg, element, cache)
+
+  # antidiffusive flux
+  calcflux_antidiffusive!(fhat1, fhat2, fstar1_L, fstar2_L, u, mesh,
+      nonconservative_terms, equations, indicator, dg, element, cache)
+
+  # Calculate volume integral contribution of low-order FV flux
+  for j in eachnode(dg), i in eachnode(dg)
+    for v in eachvariable(equations)
+      du[v, i, j, element] += inverse_weights[i] * (fstar1_L[v, i+1, j] - fstar1_R[v, i, j]) +
+                              inverse_weights[j] * (fstar2_L[v, i, j+1] - fstar2_R[v, i, j])
+
+    end
+  end
+
+  return nothing
+end
+
+
+# Calculate the DG staggered volume fluxes `fhat` in subcell FV-form inside the element
+# (**without non-conservative terms**).
+#
+# See also `flux_differencing_kernel!`.
+@inline function calcflux_fhat!(fhat1, fhat2, u,
+                                mesh::TreeMesh{2}, nonconservative_terms::False, equations,
+                                volume_flux, dg::DGSEM, element, cache)
+
+  @unpack weights, derivative_split = dg.basis
+  @unpack flux_temp_threaded = cache
+
+  flux_temp = flux_temp_threaded[Threads.threadid()]
+
+  # The FV-form fluxes are calculated in a recursive manner, i.e.:
+  # fhat_(0,1)   = w_0 * FVol_0,
+  # fhat_(j,j+1) = fhat_(j-1,j) + w_j * FVol_j,   for j=1,...,N-1,
+  # with the split form volume fluxes FVol_j = -2 * sum_i=0^N D_ji f*_(j,i).
+
+  # To use the symmetry of the `volume_flux`, the split form volume flux is precalculated
+  # like in `calc_volume_integral!` for the `VolumeIntegralFluxDifferencing`
+  # and saved in in `flux_temp`.
+
+  # Split form volume flux in orientation 1: x direction
+  flux_temp .= zero(eltype(flux_temp))
+
+  for j in eachnode(dg), i in eachnode(dg)
+    u_node = get_node_vars(u, equations, dg, i, j, element)
+
+    # All diagonal entries of `derivative_split` are zero. Thus, we can skip
+    # the computation of the diagonal terms. In addition, we use the symmetry
+    # of the `volume_flux` to save half of the possible two-point flux
+    # computations.
+    for ii in (i+1):nnodes(dg)
+      u_node_ii = get_node_vars(u, equations, dg, ii, j, element)
+      flux1 = volume_flux(u_node, u_node_ii, 1, equations)
+      multiply_add_to_node_vars!(flux_temp, derivative_split[i, ii], flux1, equations, dg, i,  j)
+      multiply_add_to_node_vars!(flux_temp, derivative_split[ii, i], flux1, equations, dg, ii, j)
+    end
+  end
```

Comment:
This is basically a copy-pasted code from another kernel, isn't it? Can you call the original method instead?

Reply:
Yes, it's from `flux_differencing_kernel!`. Unfortunately, in there the calculations are added directly to `du`. In addition, here we need the fluxes separated for both spatial directions, which is not possible in the original kernel.

Reply:
I see, thanks